File: keep-bridge/src/keepClient.ts
import type { KeepSnapshot } from "./types.js";
import { google, keep_v1 } from "googleapis";
import { OAuth2Client } from "google-auth-library";

export interface KeepClient {
  fetchSnapshot(): Promise<KeepSnapshot>;
}

export class MockKeepClient implements KeepClient {
  private counter = 0;

  async fetchSnapshot(): Promise<KeepSnapshot> {
    this.counter += 1;

    return {
      noteId: process.env.KEEP_TARGET_NOTE_ID ?? "mock-note",
      title: process.env.PRINT_TITLE ?? "Shopping List",
      uncheckedItems: ["Milk", "Eggs", "Bread"],
      checkedItems: this.counter % 2 === 0 ? ["Apples"] : [],
      updatedAt: new Date().toISOString()
    };
  }
}

export class GoogleKeepClient implements KeepClient {
  private readonly targetNoteId: string;
  private readonly printTitleFallback: string;
  private readonly oauthClient: OAuth2Client;
  private readonly keepApi: keep_v1.Keep;

  constructor() {
    const clientId = process.env.KEEP_GOOGLE_CLIENT_ID;
    const clientSecret = process.env.KEEP_GOOGLE_CLIENT_SECRET;
    const refreshToken = process.env.KEEP_GOOGLE_REFRESH_TOKEN;
    this.targetNoteId = process.env.KEEP_TARGET_NOTE_ID ?? "";
    this.printTitleFallback = process.env.PRINT_TITLE ?? "Shopping List";

    if (!clientId || !clientSecret || !refreshToken || !this.targetNoteId) {
      throw new Error(
        "Missing Keep config. Expected KEEP_GOOGLE_CLIENT_ID, KEEP_GOOGLE_CLIENT_SECRET, KEEP_GOOGLE_REFRESH_TOKEN, KEEP_TARGET_NOTE_ID"
      );
    }

    const redirectUri = process.env.KEEP_GOOGLE_REDIRECT_URI ?? "http://localhost:8787/oauth2callback";

    this.oauthClient = new OAuth2Client(clientId, clientSecret, redirectUri);
    this.oauthClient.setCredentials({ refresh_token: refreshToken });
    this.keepApi = google.keep({ version: "v1", auth: this.oauthClient });
  }

  async fetchSnapshot(): Promise<KeepSnapshot> {
    const noteName = normalizeNoteName(this.targetNoteId);
    const noteResponse = await this.keepApi.notes.get({ name: noteName });
    const note = noteResponse.data;

    const listItems = note.body?.list?.listItems ?? [];
    const uncheckedItems: string[] = [];
    const checkedItems: string[] = [];

    for (const listItem of listItems) {
      const text = extractListItemText(listItem).trim();
      if (!text) {
        continue;
      }

      if (listItem.checked) {
        checkedItems.push(text);
      } else {
        uncheckedItems.push(text);
      }
    }

    return {
      noteId: note.name ?? noteName,
      title: note.title?.trim() || this.printTitleFallback,
      uncheckedItems,
      checkedItems,
      updatedAt: note.updateTime ?? new Date().toISOString()
    };
  }
}

function normalizeNoteName(noteIdOrName: string): string {
  return noteIdOrName.startsWith("notes/") ? noteIdOrName : `notes/${noteIdOrName}`;
}

function extractListItemText(item: keep_v1.Schema$ListItem): string {
  const textField = item.text as unknown;

  if (typeof textField === "string") {
    return textField;
  }

  if (textField && typeof textField === "object") {
    if ("text" in textField && typeof (textField as { text?: unknown }).text === "string") {
      return (textField as { text: string }).text;
    }
  }

  return "";
}

export function createKeepClient(): KeepClient {
  const useMock = process.env.KEEP_USE_MOCK === "true";
  if (useMock) {
    return new MockKeepClient();
  }

  return new GoogleKeepClient();
}

File: keep-bridge/src/keep_client.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Protocol

import gkeepapi
from gkeepapi import node


@dataclass(frozen=True)
class KeepSnapshot:
    note_id: str
    title: str
    unchecked_items: list[str]
    checked_items: list[str]
    updated_at: str


class KeepClient(Protocol):
    def fetch_list(self, list_name: str) -> KeepSnapshot | None:
        ...

    def keepalive_sync(self) -> None:
        ...


class MockKeepClient:
    def __init__(self) -> None:
        self._counter = 0

    def fetch_list(self, list_name: str) -> KeepSnapshot | None:
        self._counter += 1
        normalized = list_name.strip().casefold()
        if normalized != "shopping list" and normalized != "shopping list mock":
            return None

        return KeepSnapshot(
            note_id="mock-note",
            title=list_name.strip() or "Shopping List",
            unchecked_items=["Milk", "Eggs", "Bread"],
            checked_items=["Apples"] if self._counter % 2 == 0 else [],
            updated_at=datetime.now(tz=timezone.utc).isoformat(),
        )

    def keepalive_sync(self) -> None:
        return


class GoogleKeepClient:
    def __init__(self) -> None:
        self._email = require_env("KEEP_EMAIL")
        self._master_token = require_env("KEEP_MASTER_TOKEN")
        self._state_file = Path(os.getenv("KEEP_STATE_FILE", "keep_state.json"))

        self._keep = gkeepapi.Keep()
        state = self._load_state()
        try:
            # gkeepapi.authenticate returns None on success; failures raise exceptions.
            self._keep.authenticate(self._email, self._master_token, state=state)
        except Exception as error:  # noqa: BLE001
            raise RuntimeError(
                "Failed to authenticate with gkeepapi. Check KEEP_EMAIL and KEEP_MASTER_TOKEN."
            ) from error

    def keepalive_sync(self) -> None:
        self._sync_and_persist()

    def fetch_list(self, list_name: str) -> KeepSnapshot | None:
        lookup_name = list_name.strip()
        if not lookup_name:
            return None

        self._sync_and_persist()

        target_note = self._find_list_by_title(lookup_name)
        if target_note is None:
            return None

        unchecked_items = [item.text.strip() for item in target_note.unchecked if item.text and item.text.strip()]
        checked_items = [item.text.strip() for item in target_note.checked if item.text and item.text.strip()]

        updated_ts = target_note.timestamps.updated
        updated_at = updated_ts.isoformat() if updated_ts else datetime.now(tz=timezone.utc).isoformat()

        return KeepSnapshot(
            note_id=target_note.id,
            title=target_note.title.strip() if target_note.title else lookup_name,
            unchecked_items=unchecked_items,
            checked_items=checked_items,
            updated_at=updated_at,
        )

    def _sync_and_persist(self) -> None:
        self._keep.sync()
        self._save_state()

    def _find_list_by_title(self, title: str) -> node.List | None:
        title_folded = title.casefold()
        exact_matches: list[node.List] = []
        fuzzy_matches: list[node.List] = []

        for n in self._keep.all():
            if not isinstance(n, node.List):
                continue
            if n.trashed or n.deleted:
                continue

            note_title = (n.title or "").strip()
            note_title_folded = note_title.casefold()

            if note_title_folded == title_folded:
                exact_matches.append(n)
            elif title_folded in note_title_folded:
                fuzzy_matches.append(n)

        if exact_matches:
            return exact_matches[0]
        if fuzzy_matches:
            return fuzzy_matches[0]
        return None

    def _load_state(self) -> dict | None:
        if not self._state_file.exists():
            return None

        try:
            with self._state_file.open("r", encoding="utf-8") as fh:
                return json.load(fh)
        except (json.JSONDecodeError, OSError):
            return None

    def _save_state(self) -> None:
        state = self._keep.dump()
        self._state_file.parent.mkdir(parents=True, exist_ok=True)
        with self._state_file.open("w", encoding="utf-8") as fh:
            json.dump(state, fh)


def create_keep_client() -> KeepClient:
    use_mock = os.getenv("KEEP_USE_MOCK", "false").lower() == "true"
    if use_mock:
        return MockKeepClient()
    return GoogleKeepClient()


def require_env(key: str) -> str:
    value = os.getenv(key)
    if not value:
        raise RuntimeError(f"Missing required environment variable: {key}")
    return value
